File: solutions/clojure/atbash-cipher/1/src/atbash_cipher.clj
(ns atbash-cipher)


(def abc "abcdefghijklmnopqrstuvwxyz")

(def cipher (zipmap abc (reverse abc)))


(defn encode [plain]
  (->> plain
       clojure.string/lower-case
       (filter #(Character/isLetterOrDigit %))
       (map #(cipher % %))
       (partition-all 5)
       (map (partial apply str))
       (clojure.string/join " ")))

File: solutions/python/atbash-cipher/1/atbash_cipher.py
"""Encode and decode text using the Atbash substitution cipher."""

from string import ascii_lowercase, digits


TRANSLATION = str.maketrans(ascii_lowercase, ascii_lowercase[::-1])
ALPHANUMERIC = frozenset(ascii_lowercase + digits)


def encode(plain_text: str) -> str:
    """Encode text in groups of five letters or digits."""
    cipher = decode(plain_text)
    return " ".join(cipher[index:index + 5] for index in range(0, len(cipher), 5))


def decode(ciphered_text: str) -> str:
    """Reverse the alphabet substitution and discard non-alphanumeric characters."""
    cleaned = "".join(char for char in ciphered_text.lower() if char in ALPHANUMERIC)
    return cleaned.translate(TRANSLATION)
